src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,913 @@
+                                           (sym (ℕp.+-assoc 1 m (1+ n)))
+                                    ⟩
+  2 *ₙ ((1+ m) +ₙ (1+ n))          ≡⟨ ℕp.*-distribˡ-+ 2 (1+ m) (1+ n) ⟩
+  (2 *ₙ (1+ m)) +ₙ (2 *ₙ (1+ n))   ≡⟨⟩
+  toℕ (2suc x) +ₙ toℕ (2suc y)   ∎
+  where
+  m = toℕ x;  n = toℕ y
+
+toℕ+homo (2suc x) (suc2* y) = begin
+  toℕ ((2suc x) + (suc2* y))      ≡⟨⟩
+  toℕ (suc (2suc (x + y)))        ≡⟨ toℕ∘suc (2suc (x + y)) ⟩
+  1+ (toℕ (2suc (x + y)))         ≡⟨⟩
+  1+ (2 *ₙ (1+ (toℕ (x + y))))    ≡⟨ cong (1+_ ∘ (2 *ₙ_) ∘ 1+_) (toℕ+homo x y)
+                                   ⟩
+  1+ (2 *ₙ (1+ (m +ₙ n)))         ≡⟨ cong (1+_ ∘ (2 *ₙ_))
+                                                 (sym (ℕp.+-assoc 1 m n)) ⟩
+  1+ (2 *ₙ (1+m +ₙ n))            ≡⟨ cong 1+_ (ℕp.*-distribˡ-+ 2 1+m n) ⟩
+  1+ ((2 *ₙ 1+m) +ₙ 2 *ₙ n)       ≡⟨ Of+ℕ-semigroup.x∙yz≈y∙xz 1 _ (2 *ₙ n) ⟩
+  (2 *ₙ 1+m) +ₙ (1+ (2 *ₙ n))     ≡⟨⟩
+  toℕ (2suc x) +ₙ toℕ (suc2* y)   ∎
+  where
+  m = toℕ x;  n = toℕ y;  1+m = 1+ m
+
+toℕ+homo (suc2* x) (2suc y) = begin
+  toℕ ((suc2* x) + (2suc y))      ≡⟨⟩
+  toℕ (suc (2suc (x + y)))        ≡⟨ toℕ∘suc (2suc (x + y)) ⟩
+  1+ (toℕ (2suc (x + y)))         ≡⟨⟩
+  1+ (2 *ₙ (1+ (toℕ (x + y))))    ≡⟨ cong (1+_ ∘ (2 *ₙ_) ∘ 1+_)
+                                                         (toℕ+homo x y) ⟩
+  1+ (2 *ₙ (1+ (m +ₙ n)))         ≡⟨ cong (1+_ ∘ (2 *ₙ_))
+                                          (Of+ℕ-semigroup.x∙yz≈y∙xz 1 m n)
+                                   ⟩
+  1+ (2 *ₙ (m +ₙ 1+n))            ≡⟨ cong 1+_ (ℕp.*-distribˡ-+ 2 m 1+n) ⟩
+  1+ (2 *ₙ m +ₙ 2 *ₙ 1+n)         ≡⟨ sym (ℕp.+-assoc 1 (2 *ₙ m) (2 *ₙ 1+n)) ⟩
+  (1+ (2 *ₙ m)) +ₙ (2 *ₙ (1+ n))  ≡⟨⟩
+  toℕ (suc2* x) +ₙ toℕ (2suc y)   ∎
+  where
+  m = toℕ x;  n = toℕ y;  1+n = 1+ n
+
+toℕ+homo (suc2* x) (suc2* y) = begin
+  toℕ ((suc2* x) + (suc2* y))       ≡⟨⟩
+  toℕ (suc (suc2* (x + y)))         ≡⟨ toℕ∘suc (suc2* (x + y)) ⟩
+  1+ (toℕ (suc2* (x + y)))          ≡⟨⟩
+  1+ (1+ (2 *ₙ (toℕ (x + y))))      ≡⟨ cong (1+_ ∘ 1+_ ∘ (2 *ₙ_))
+                                                       (toℕ+homo x y) ⟩
+  1+ (1+ (2 *ₙ (m +ₙ n)))           ≡⟨ cong (1+_ ∘ 1+_) (ℕp.*-distribˡ-+ 2 m n)
+                                     ⟩
+  1+ (1+ (2 *ₙ m +ₙ 2 *ₙ n))        ≡⟨ cong 1+_
+                                            (sym (ℕp.+-assoc 1 (2 *ₙ m) _)) ⟩
+  1+ ((1+ (2 *ₙ m) +ₙ (2 *ₙ n)))    ≡⟨ (Of+ℕ-semigroup.x∙yz≈y∙xz
+                                                         1 (1+ (2 *ₙ m)) _) ⟩
+  (1+ (2 *ₙ m)) +ₙ (1+ (2 *ₙ n))    ≡⟨⟩
+  toℕ (suc2* x) +ₙ toℕ (suc2* y)    ∎
+  where
+  m = toℕ x;  n = toℕ y
+
+
+ext-fromℕ :  (n : ℕ) → ∃ (\x → toℕ x ≡ n)    -- Mind:  it costs O(n)
+ext-fromℕ 0       =  (0# , refl)
+ext-fromℕ (1+ n)  with ext-fromℕ n
+...
+    | (x , toℕ-x≡n) = (suc x , toℕ-suc-x≡1+n)
+  where
+  toℕ-suc-x≡1+n = begin
+    toℕ (suc x)   ≡⟨ toℕ∘suc x ⟩
+    1+ (toℕ x)    ≡⟨ cong 1+_ toℕ-x≡n ⟩
+    1+ n          ∎
+
+fromℕ : ℕ → Bin
+fromℕ = proj₁ ∘ ext-fromℕ
```

Comment:
`fromℕ` should be in `Data.Bin.Base` not in `Data.Bin.Properties`. The proof `ext-fromℕ` should stay here though (if still needed which I'm not sure about).

Reply:
How can it move there? 
1) ``fromℕ`` is defined as ``proj₁ ∘ ext-fromℕ``, 
2) ``ext-fromℕ`` uses the lemma ``toℕ∘suc`` from Bin.Properties,
   moving ``fromℕ`` to Bin.Base will lead to import cycle.

> [..] (if still needed which I'm not sure about).

Also 
a) ``ext-fromℕ`` is used below in ``fromℕ∘toℕ, fromℕ∘pred``,
b) it actually proves that fromℕ is a left inverse to ``toℕ``.

Generally: without separating proofs from ordinary evaluation we cannot separate the modules of kind ``Properties`` from functions computing ordinary values.

And if we separate proofs from ordinary evaluation (consider DivMod, GCD ...),
then we shall actually have to program the corresponding functions twice: first to program ordinary computation, then to program proof construction by applying a similar recursions (with repeating the environment import).

Reply:
> How can it move there?

By writing `fromℕ` as:
```agda
fromℕ :  (n : ℕ) → ∃ (\x → toℕ x ≡ n)`
fromℕ zero   = 0#
fromℕ (1+ n) = suc n
```

> And if we separate proofs from ordinary evaluation (consider DivMod, GCD ...),
then we shall actually have to program the corresponding functions twice: first to program ordinary computation, then to program proof construction by applying a similar recursions (with repeating the environment import).

In general the standard library takes the view that separating the proofs from evaluation is desirable. It:
1. makes it clearer how the function is evaluated.
2. in many cases the function evaluation is fast, but the building the proofs is slow. Therefore combining both often slows down the performance of evaluation drastically. For example that's exactly why `GCD` was so slow and why #770 which separated evaluation from the proofs speeds it up.